### PR TITLE
MAINT: fix usage of deprecated Numpy C API in signal

### PR DESCRIPTION
**Pull request by rgommers**

### CODE REVIEW

File: scipy/signal/sigtoolsmodule.c [hunk 1]
```diff
@@ -762,15 +770,17 @@ static int pre_remez(double *h2, int numtaps, int numbands, double *bands, doubl
...
-  intp *dims2 = ap2->dimensions;
-  intp *dims1 = ap1->dimensions;
-  intp is1 = ap1->descr->elsize;
-  intp is2 = ap2->descr->elsize;
-  char *ip2 = ap2->data;
-  int elsize = ap1->descr->elsize;
+  int ndims = PyArray_NDIM(ap1);
+  intp *dims2 = PyArray_DIMS(ap2);
+  intp *dims1 = PyArray_DIMS(ap1);
+  intp is1 = PyArray_DESCR(ap1)->elsize;
```
Comment (WarrenWeckesser) [inline scipy/signal/sigtoolsmodule.c]: This (and other references to `->descr->elsize` elsewhere) could be written 

    intp is1 = PyArray_ITEMSIZE(ap1);

File: scipy/signal/sigtoolsmodule.c [hunk 2]
```diff
@@ -856,51 +867,56 @@ PyObject *PyArray_OrderFilterND(PyObject *op1, PyObject *op2, int order) {
...
 	}
 
 	n2 = PyArray_Size((PyObject *)ap2);
 	n2_nonzero = 0;
-	ap2_ptr = ap2->data;
-	/* Find out the number of non-zero entries in domain (allows for
-	 *  different shapped rank-filters to be used besides just rectangles)
+	ap2_ptr = PyArray_DATA(ap2);
+	/* 
+         * Find out the number of non-zero entries in domain (allows for
```
Comment (WarrenWeckesser) [inline scipy/signal/sigtoolsmodule.c]: Extra space before `*`

File: scipy/signal/sigtoolsmodule.c [hunk 3]
```diff
@@ -934,22 +952,27 @@ PyObject *PyArray_OrderFilterND(PyObject *op1, PyObject *op2, int order) {
...
+	for (k=0; k < PyArray_NDIM(ap1); k++) {
+            a_ind[k] = mode_dep[k];
+            check_ind[k] = PyArray_DIMS(ap1)[k] - PyArray_DIMS(ap2)[k] - mode_dep[k] - 1;
+        }
+	a_ind[PyArray_NDIM(ap1)-1]--;
 	i = PyArray_Size((PyObject *)ret);
 	while (i--) {
-	  /* Zero out the sort_buffer (has effect of zero-padding
-	     on boundaries). Treat object arrays right.*/
+	  /*
```
Comment (WarrenWeckesser) [inline scipy/signal/sigtoolsmodule.c]: Align `*`s that follow with this one.

File: scipy/signal/sigtoolsmodule.c [hunk 4]
```diff
@@ -1037,22 +1066,28 @@ static PyObject *sigtools_convolve2d(PyObject *NPY_UNUSED(dummy), PyObject *args
...
     switch(mode & OUTSIZE_MASK) {
     case VALID:
-	for (i = 0; i < ain1->nd; i++) { 
-	    aout_dimens[i] = ain1->dimensions[i] - ain2->dimensions[i] + 1;
+	for (i = 0; i < PyArray_NDIM(ain1); i++) { 
+	    aout_dimens[i] = PyArray_DIMS(ain1)[i] - PyArray_DIMS(ain2)[i] + 1;
 	    if (aout_dimens[i] < 0) {
-		PyErr_SetString(PyExc_ValueError, "no part of the output is valid, use option 1 (same) or 2 (full) for third argument");
+		PyErr_SetString(PyExc_ValueError,
+                    "no part of the output is valid, use option 1 (same) or 2 " \
```
Comment (WarrenWeckesser) [inline scipy/signal/sigtoolsmodule.c]: The trailing `\` isn't necessary--C has string literal concatenation.

**Comment by juliantaylor** [comment]
are there more deprecated uses in scipy? instead of doing it by hand it seems these type of changes can be handled automatically with a coccinelle patch: http://coccinelle.lip6.fr/

**Comment by juliantaylor** [comment]
urg a patch handling only three cases alreade produces a 438 insertions(+), 396 deletions(-) diff :/

```
@@ PyArrayObject *arr; @@
- arr->nd
+ PyArray_NDIM(arr)

@@ PyArrayObject *arr; @@
- arr->descr
+ PyArray_DESCR(arr)

@@ PyArrayObject *arr; @@
- arr->dimensions
+ PyArray_DIMS(arr)
```

**Comment by WarrenWeckesser** [comment]
@juliantaylor coccinelle is pretty slick!  Thanks for the pointer.  I'm experimenting with it now in scipy/integrate.

**Comment by rgommers** [comment]
@juliantaylor looks interesting. I did this with some Vim macros but checked all the changes manually. There are some gotchas, like ``xxx->strides`` needing to be changed to ``PyArray_STRIDES`` only if ``xxx`` is an array and not if it's an iterator. Would coccinelle be able to handle that kind of thing?

**Comment by juliantaylor** [comment]
yes coccinelle is type aware and knows the C invariants, it can replace only those xxx->strides where xxx is of a certain type, it is incredibly powerful (see the many examples) but for non trivial transformations it has a steep learning curve. But these simple member access -> function call replacements are easy, you run the patch I posted above with:

    spatch --sp-file file.patch --in-place file
    # for the full tree with gnu parallel:
    find . -name "*.c*" | parallel spatch --in-place --sp-file file.spatch

**Comment by rgommers** [comment]
@juliantaylor thanks! I'll play with it

File: scipy/signal/sigtoolsmodule.c (hunk 3 above)
Comment (rgommers) [inline scipy/signal/sigtoolsmodule.c]: hmm, both this one and the one above are tabs-vs-spaces. I'll fix things in these two places but am not going to attempt global replace of tabs.

**Comment by rgommers** [comment]
Looks like we'll do the rest of Scipy with Coccinelle, but since this was pretty much done I just fixed the few issues pointed out by Warren.

File: scipy/signal/splinemodule.c [hunk 5]
```diff
@@ -4,14 +4,10 @@
...
-#define STRIDES(arr) ((arr)->strides)
-#define ELSIZE(arr) ((arr)->descr->elsize)
-#define OBJECTTYPE(arr) ((arr)->descr->type_num)
-#define BASEOBJ(arr) ((PyArrayObject *)((arr)->base))
-#define RANK(arr) ((arr)->nd)
-#define ISCONTIGUOUS(m) ((m)->flags & NPY_CONTIGUOUS)
+#define DATA(arr) (PyArray_DATA(arr))
+#define DIMS(arr) (PyArray_DIMS(arr))
+#define STRIDES(arr) (PyArray_STRIDES(arr))
+#define ELSIZE(arr) (PyArray_ITEMSIZE(arr))
```
Comment (larsmans) [inline scipy/signal/splinemodule.c]: I'd get rid rid of these last four macros too and use the NumPy ones.

Comment (WarrenWeckesser) [inline scipy/signal/splinemodule.c]: +1

Comment (rgommers) [inline scipy/signal/splinemodule.c]: okay will do

**Comment by rgommers** [comment]
TravisCI failure was one timeout - everything else is green. Should be good to go.

**Comment by rgommers** [comment]
Going to merge this soon unless there are more comments.

**Comment by rgommers** [comment]
merged. thanks all for the feedback

**Comment by WarrenWeckesser** [comment]
Thanks, Ralf.

**Comment by juliantaylor** [comment]
have you made some changes to the coccinelle patch? I wanted to give it some more polish too, maybe a shared gist would be useful

**Comment by rgommers** [comment]
@juliantaylor not yet. This PR I still did the old-fashioned way. 

For the other modules, I plan to do that by a single Coccinelle patch, one module at a time. I think one commit for each Coccinelle change and then a second commit to clean up indentation, break lines to 80 chars, etc. And if needed fix build warnings - I noticed that macros like ``PyArray_DATA`` give new warnings because parameters are declared as ``const`` - needs extra casts.

**Comment by WarrenWeckesser** [comment]
@juliantaylor: I experimented a bit with this:

```
@@ PyArrayObject *arr; @@
- (arr)->nd
+ PyArray_NDIM(arr)

@@ PyArrayObject *arr; @@
- (arr)->dimensions
+ PyArray_DIMS(arr)

@@ PyArrayObject *arr; @@
- (arr)->data
+ PyArray_DATA(arr)

@@ PyArrayObject *arr; @@
- (arr)->strides
+ PyArray_STRIDES(arr)

@@ PyArrayObject *arr; @@
- (arr)->flags
+ PyArray_FLAGS(arr)

@@ PyArrayObject *arr; @@
- (arr)->descr->elsize
+ PyArray_ITEMSIZE(arr)

@@ PyArrayObject *arr; @@
- (arr)->descr
+ PyArray_DESCR(arr)
```
For this file:
```
int func(int k)
{
    PyArrayObject *p;
    PyArrayObject **q;
    npy_intp *strides;
    npy_intp *dims;
    npy_intp ndim;
    int flags;
    double *data;

    data = p->data;

    data = (p)->data;

    data = (*q)->data;

    data = (*(q))->data;

    k = p->descr->elsize;

    flags = p->flags | 1;

    strides = p->strides;

    ndim = p->nd;

    k = p->dimensions[0];

    p->descr;

    return k & (*q)->flags;
}
```
the diff looks OK:
```
--- tmp.c
+++ /var/folders/p9/j_v_3wvd2bx5ftq544pvvf1w0000gn/T/cocci-output-24538-a96123-tmp.c
@@ -9,25 +9,25 @@ int func(int k)
     int flags;
     double *data;
 
-    data = p->data;
+    data = PyArray_DATA(p);
 
-    data = (p)->data;
+    data = PyArray_DATA(p);
 
-    data = (*q)->data;
+    data = PyArray_DATA(*q);
 
-    data = (*(q))->data;
+    data = PyArray_DATA(*(q));
 
-    k = p->descr->elsize;
+    k = PyArray_ITEMSIZE(p);
 
-    flags = p->flags | 1;
+    flags = PyArray_FLAGS(p) | 1;
 
-    strides = p->strides;
+    strides = PyArray_STRIDES(p);
 
-    ndim = p->nd;
+    ndim = PyArray_NDIM(p);
 
-    k = p->dimensions[0];
+    k = PyArray_DIMS(p)[0];
 
-    p->descr;
+    PyArray_DESCR(p);
 
-    return k & (*q)->flags;
+    return k & PyArray_FLAGS(*q);
 }
```

**Comment by WarrenWeckesser** [comment]
Ralf, I've started making similar changes in `scipy.integrate`, but I don't want to duplicate any work you might have in progress.  Are you already working on a similar change for `scipy.integrate`?

**Comment by rgommers** [comment]
No, I haven't done much more of this on any module recently. Please go ahead.

**Pull request merged**